keyring.go

```diff
@@ -0,0 +1,303 @@
+package gomatrixserverlib
+
+import (
+	"fmt"
+	"golang.org/x/crypto/ed25519"
+	"strings"
+	"time"
+)
+
+// A Timestamp is a millisecond posix timestamp.
+type Timestamp uint64
+
+// A PublicKeyRequest is a request for a public key with a particular key ID.
+type PublicKeyRequest struct {
+	// The server to fetch a key for.
+	ServerName string
+	// The ID of the key to fetch.
+	KeyID string
+}
+
+// A KeyFetcher is a way of fetching public keys in bulk.
+type KeyFetcher interface {
+	// Lookup a batch of public keys.
+	FetchKeys(requests map[PublicKeyRequest]Timestamp) (map[PublicKeyRequest]ServerKeys, error)
+}
+
+// A KeyDatabase is a store for caching public keys.
+type KeyDatabase interface {
+	KeyFetcher
+	// Add a block public keys to the database.
```

Comment:
Done 6251765. I don't know if one follows the `<name of method>... does` for interfaces.